odoo/addons/base/models/ir_model.py

```diff
@@ -1319,6 +1319,86 @@ def get_field_selection(self, model_name, field_name):
         field = self._get(model_name, field_name)
         return [(sel.value, sel.name) for sel in field.selection_ids]
 
+
+class ModelInherit(models.Model):
+    _name = "ir.model.inherit"
+    _description = "Model Inheritance Tree"
+
+    model_id = fields.Many2one("ir.model", required=True, ondelete="cascade")
+    inherit_id = fields.Many2one("ir.model", required=True, ondelete="cascade")
+    field_id = fields.Many2one("ir.model.fields", ondelete="cascade")  # in case of inherits
+
+    _sql_constraints = [
+        ("uniq", "UNIQUE(model_id, inherit_id)", "")
+    ]
+
+    def _reflect_inherits(self, model_names):
+        """ Reflect the given models' inherits. """
+        get_model = self.env["ir.model"]._get_id
+        inherits = []
+
+        by_module = defaultdict(list)
+        for model_name in model_names:
+            get_field = self.env["ir.model.fields"]._get_ids(model_name).get
+            model_id = get_model(model_name)
+            model = self.env[model_name]
+
+            for cls in reversed(type(model).mro()):
+                if not models.is_definition_class(cls):
+                    continue
+
+                inherits = [
+                    (model_id, get_model(inherit), None) for inherit in cls._inherit
+                    if inherit != "base"
+                    if inherit != model_name
+                ]
+                inherits += [
+                    (model_id, get_model(inherit), get_field(field))
+                    for inherit, field in cls._inherits.items()
+                ]
+
+                for inh in inherits:
+                    by_module[inh].append(cls._module)
+
+        if not by_module:
+            return
+
+        cr = self.env.cr
+        cr.execute(
+            """
+                SELECT i.id, i.model_id, i.inherit_id, i.field_id
+                  FROM ir_model_inherit i
+                  JOIN ir_model m
+                    ON m.id = i.model_id
+                 WHERE m.model IN %s
+            """,
+            [tuple(model_names)]
+        )
+        existing = {}
+        inh_ids = {}
+        for iid, model_id, inherit_id, field_id in cr.fetchall():
+            inh_ids[(model_id, inherit_id, field_id)] = iid
+            existing[(model_id, inherit_id)] = field_id
+
+        sentinel = object()
+        cols = ["model_id", "inherit_id", "field_id"]
+        rows = [inh for inh in by_module if existing.get(inh[:2], sentinel) != inh[2]]
+        if rows:
+            ids = upsert_en(self, cols, rows, ["model_id", "inherit_id"])
+            for row, id_ in zip(rows, ids):
+                inh_ids[row] = id_
+            self.pool.post_init(mark_modified, self.browse(ids), cols[1:])
+
+        # update their XML id
+        data_list = [
+            {"xml_id": f"{module}.model_inherit__{inh[0]}_{inh[1]}", "record": self.browse(inh_ids[inh])}
```

Comment:
Please factor out the external id definition in a function like `model_xmlid` and `field_xmlid`.
This code generates external ids that contains ids (instead of model names), which is mostly useless.